sbin/reboot/reboot.8
.\" Copyright (c) 1990, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)reboot.8	8.1 (Berkeley) 6/9/93
.\"
.Dd June 9, 1993
.Dt REBOOT 8
.Os
.Sh NAME
.Nm reboot ,
.Nm halt
.Nd
stopping and restarting the system
.Sh SYNOPSIS
.Nm halt
.Op Fl dhnq
.Nm reboot
.Op Fl dnq

.Nm fasthalt
.Nm fastboot
.Sh DESCRIPTION
The
.Nm halt
and
.Nm reboot
utilities flush the filesystem cache to disk, send all running processes
a SIGTERM (and subsequently a SIGKILL) and, respectively, halt or restart
the system.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl d
The
.Fl d
option causes the system to make a
.Dq crash dump
(debugging image)
when halting or rebooting.
.It Fl h
Once the system has halted, attempt to power down the machine.
If the hardware does not support this feature this option
is silently ignored.
.It Fl n
If the
.Fl n
option is specified,
the filesystem cache is not flushed.
This option should probably not be used.
.It Fl q
If the
.Fl q
option is specified,
the system is halted or restarted quickly and ungracefully,
and only the flushing of the filesystem cache is performed.
This option should probably not be used.
.El
.Pp
The reboot or halt action is logged,
including entering a shutdown record into the login
accounting file; see
.Xr syslogd 8
and
.Xr wtmp 5
for more information.
.Pp
Normally, the
.Xr shutdown 8
utility is used when the system needs to be halted or restarted,
giving users advance warning of their impending doom.
.Sh SEE ALSO
.Xr wtmp 5 ,
.Xr boot 8 ,
.Xr shutdown 8 ,
.Xr sync 8
.Sh COMPATIBILITY
Historically, the
.Nm fastboot
and
.Nm fasthalt
commands were shell scripts that arranged for the filesystems not to be
checked on reboot.
As filesystems are no longer checked unless necessary,
they have been deleted and are now simply links to the reboot and halt
commands.
.Sh HISTORY
A
.Nm reboot
command appeared in
.At v6 .


sbin/reboot/reboot.c
/*
 * Copyright (c) 1980, 1986, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
static char copyright[] =
"@(#) Copyright (c) 1980, 1986, 1993\n\
	The Regents of the University of California.  All rights reserved.\n";
#endif /* not lint */

#ifndef lint
static char sccsid[] = "@(#)reboot.c	8.1 (Berkeley) 6/5/93";
#endif /* not lint */

#include <sys/types.h>
#include <sys/reboot.h>

#include <err.h>
#include <errno.h>
#include <pwd.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <unistd.h>

void usage __P((void));

int dohalt;

int
main(argc, argv)
	int argc;
	char *argv[];
{
	struct passwd *pw;
	int ch, howto, i, lflag, nflag, qflag;
	char *p, *user;

	if ((p = strrchr(*argv, '/')) == NULL)
		p = *argv;
	else
		++p;
	if (*p == '-')
		p++;
	if (strcmp(p, "halt") == 0 || strcmp(p, "fasthalt") == 0) {
		dohalt = 1;
		howto = RB_HALT;
	} else
		howto = 0;

	lflag = nflag = qflag = 0;
	while ((ch = getopt(argc, argv, "dhlnq")) != EOF)
		switch (ch) {
		case 'd':
			howto |= RB_DUMP;
			break;
		case 'h':
			howto |= RB_POWEROFF;
			break;
		case 'l':		/* Undocumented; used by shutdown. */
			lflag = 1;
			break;
		case 'n':
			nflag = 1;
			howto |= RB_NOSYNC;
			break;
		case 'q':
			qflag = 1;
			break;
		case '?':
		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	if (geteuid())
		errx(1, "%s", strerror(EPERM));

	if (qflag) {
		reboot(howto);
		err(1, NULL);
	}

	/* Log the reboot. */
	if (!lflag)  {
		if ((user = getlogin()) == NULL)
			user = (pw = getpwuid(getuid())) ? pw->pw_name : "???";
		if (dohalt) {
			openlog("halt", 0, LOG_AUTH | LOG_CONS);
			syslog(LOG_CRIT, "halted by %s", user);
		} else {
			openlog("reboot", 0, LOG_AUTH | LOG_CONS);
			syslog(LOG_CRIT, "rebooted by %s", user);
		}
	}
	logwtmp("~", "shutdown", "");

	/*
	 * Do a sync early on, so disks start transfers while we're off
	 * killing processes.  Don't worry about writes done before the
	 * processes die, the reboot system call syncs the disks.
	 */
	if (!nflag)
		sync();

	/* Just stop init -- if we fail, we'll restart it. */
	if (kill(1, SIGTSTP) == -1)
		err(1, "SIGTSTP init");

	/* Ignore the SIGHUP we get when our parent shell dies. */
	(void)signal(SIGHUP, SIG_IGN);

	/* Send a SIGTERM first, a chance to save the buffers. */
	if (kill(-1, SIGTERM) == -1)
		err(1, "SIGTERM processes");

	/*
	 * After the processes receive the signal, start the rest of the
	 * buffers on their way.  Wait 5 seconds between the SIGTERM and
	 * the SIGKILL to give everybody a chance.
	 */
	sleep(2);
	if (!nflag)
		sync();
	sleep(3);

	for (i = 1;; ++i) {
		if (kill(-1, SIGKILL) == -1) {
			if (errno == ESRCH)
				break;
			warn("killing processes");
			goto restart;
		}
		if (i > 5) {
			warnx("WARNING: some process(es) wouldn't die!");
			break;
		}
		(void)sleep(2 * i);
	}

	reboot(howto);
	warn("reboot");

	/* FALLTHROUGH */

restart:
	if (kill(1, SIGHUP) == -1)
		warn("unable to restart init");
	return (1);
}

void
usage()
{
	(void)fprintf(stderr, "usage: %s [-dhnq]\n", dohalt ? "halt" : "reboot");
	exit(1);
}
